File: Problems/Algorithms/200. Number of Islands/NumberOfIslands.kt
class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        val n = grid.size
        val m = grid[0].size
        val visited = Array(n) { BooleanArray(m) { false } }
        val neighbors = arrayOf(intArrayOf(0, 1), intArrayOf(0, -1), intArrayOf(1, 0), intArrayOf(-1, 0))
        
        var ans = 0
        for (i in 0..n-1) {
            for (j in 0..m-1) {
                if (!visited[i][j] && grid[i][j] == '1') {
                    ans++
                    val queue = ArrayDeque<IntArray>()
                    queue.add(intArrayOf(i, j))
                    
                    while (!queue.isEmpty()) {
                        for (k in 0..queue.size-1) {
                            val curr = queue.removeLast()
                            val row = curr[0]
                            val col = curr[1]
                        
                            visited[row][col] = true
                            for (neighbor in neighbors) {
                                mark(queue, row+neighbor[0], col+neighbor[1], visited, grid)
                            }
                        }
                    }
                }
            }
        }
        
        return ans
    }
    
    private fun mark(queue: ArrayDeque<IntArray>, i: Int, j: Int, visited: Array<BooleanArray>, grid: Array<CharArray>): Unit {
        if (i < 0 || i >= grid.size || j < 0 || j >= grid[0].size || grid[i][j] == '0' || visited[i][j]) return
        
        visited[i][j] = true;
        queue.add(intArrayOf(i, j))
    }
}

File: Problems/Algorithms/200. Number of Islands/number_of_islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        n, m = len(grid), len(grid[0])
        visited = [[False for _ in range(m)] for _ in range(n)]
        neighbors = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        ans = 0
        for i in range(n):
            for j in range(m):
                if not visited[i][j] and grid[i][j] == '1':
                    ans += 1

                    queue = []
                    queue.append([i, j])

                    while queue:
                        for _ in range(len(queue)):
                            row, col = queue.pop(0)
                            for neighbor in neighbors:
                                x, y = row + neighbor[0], col + neighbor[1]
                                if x >= 0 and y >= 0 and x < n and y < m and grid[x][y] == '1' and not visited[x][y]:
                                    visited[x][y] = True
                                    queue.append([x, y])

        return ans

